server/src/main/java/org/elasticsearch/index/mapper/ObjectMapper.java

```diff
@@ -553,22 +557,13 @@ private static Map<String, Mapper> buildMergedMappers(
                     // If we're merging template mappings when creating an index, then a field definition always
                     // replaces an existing one.
                     if (reason == MergeReason.INDEX_TEMPLATE) {
-                        merged = mergeWithMapper;
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeWithMapper);
                     } else {
-                        merged = mergeIntoMapper.merge(mergeWithMapper, objectMergeContext);
+                        mergedMappers.put(mergeWithMapper.simpleName(), mergeIntoMapper.merge(mergeWithMapper, objectMergeContext));
                     }
                 }
-                if (mergedMappers == null) {
-                    mergedMappers = new HashMap<>(existing.mappers);
-                }
-                mergedMappers.put(merged.simpleName(), merged);
```

Comment:
I don't think we can do that. See also https://github.com/elastic/elasticsearch/pull/102936#discussion_r1458591888

It's important to decrement the field budget before calling merge on the sub fields. I think we should have the put calls as close as possible to the place where we decrement the field budget.

Reply:
Looking at the latest changes in #102936, that no longer touches these lines, I still don't follow :)

The put is still within the loop, even if after the conditional instead of within the conditional . The only merge call I see in this method is performed within the conditional, before the put (before and after your change). What am I missing that makes this part of the changes necessary?

Reply:
I've actually reverted this in https://github.com/elastic/elasticsearch/pull/102936 now (see [c1b4258](https://github.com/elastic/elasticsearch/pull/102936/commits/c1b4258613cd5366b6d699b342e670112080fab4)). Now that I've removed the `Consumer` methods from `MapperMergeContext` and there's only `decrementFieldBudgetIfPossible` which doesn't directly add the merged mapper to the map it think it's viable to not have the put in the if/else.

What's still important though is that we decrement the budget for the shallow object mapper in `org.elasticsearch.index.mapper.ObjectMapper.MergeResult#truncateObjectMapper` before calling `shallowObjectMapper.merge`. Otherwise, if we merged the object before decrementing the budget, we wouldn't have capacity to add the object itself.